### PR TITLE
Document generation

### PR DESCRIPTION
**Pull request by johanwiden**
Make document generation work, both for C++ files and python files.
In docs/Doxyfile changed field INPUT to list relevant directories.
Added support for document generation for python files with docs/doxypypy.py and docs/py_filter.
In docs/Doxyfile updated field FILTER_PATTERNS.

A patched version of doxypypy.py is currently necessary, because Python 3.8 introduces new syntax.
This is documented in README.md and Doxyfile.

This is currently only tested under Linux. Support for Windows has been added and is documented in README.md, but this is currently untested. What could happen is that windows does not find py_filter.bat or doxypypy.py.

### CODE REVIEW

**Comment by dkeeney** [comment]
This is a great idea.  In all of our work we completely forgot about document generation.

I would prefer that it not be necessary to make an edit to use Windows.  @breznak  should we include docs in our PR build?

**Comment by breznak** [comment]
> I would prefer that it not be necessary to make an edit to use Windows. @breznak should we include docs in our PR build? 

I probably wouldn't include docs (as those generated from docs in code) to git, but it'd be nice to include in the build pipeline in CI to generate the nice-format docs and publish them somehow. If that's what you meant?

File: docs/Doxyfile_common [hunk 1]
```diff
@@ -755,42 +755,48 @@ WARN_LOGFILE           =
...
-                         ../src/nupic/math/Set.hpp \
-                         ../src/nupic/math/SparseMatrixAlgorithms.hpp \
-                         ../src/nupic/math/SparseMatrixConnections.hpp \
-                         ../src/nupic/math/SparseRLEMatrix.hpp \
-                         ../src/nupic/math/SparseTensor.hpp \
-                         ../src/nupic/math/Topology.hpp \
-                         ../src/nupic/engine/Link.hpp \
-                         ../src/nupic/engine/Output.hpp \
-                         ../src/nupic/engine/Input.hpp
+                         ../src/htm/encoders \
```
Comment (breznak) [inline docs/Doxyfile_common:758]: ok as-is, but isn't there some regex or something to look for all necessary files within a path (`../src/`) ?

**Comment by johanwiden** [comment]
The documents are now generated also under Windows.
The Doxyfile has been split up into a file for Windows and a file for Linux. These files include a
common Doxyfile plus an OS specific Doxyfile.
Also corrected a couple of paths in field INPUT in Doxyfile_common.

File: docs/Doxyfile_include_for_windows [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+# basis. Doxygen will compare the file name with each pattern and apply the
+# filter if there is a match. The filters are a list of the form: pattern=filter
+# (like *.cpp=my_cpp_filter). See INPUT_FILTER for further information on how
+# filters are used. If the FILTER_PATTERNS tag is empty or if none of the
+# patterns match the file name, INPUT_FILTER is applied.
+#
+# Note that we currently have to use a patched version of doxypypy.py because of
+# new syntax in python version 3.8.
+# See https://github.com/Feneric/doxypypy/issues/70
+# The patched version is stored in the current directory.
```
Comment (breznak) [inline docs/Doxyfile_include_for_windows:14]: how long till this is upstreamed? I'd rather not have patches-in-progress for dependencies commited directly here. Is it a big problem to say that Windows & Py3.8 are currently unsupported? When Fixed upstream, it'll automatically start to work

File: docs/Doxyfile_include_for_windows [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+FILTER_PATTERNS      = *.py=py_filter
```
Comment (breznak) [inline docs/Doxyfile_include_for_windows:16]: these files (linux, windows) are the same. I'd prefer if we can keep a single, platform independent version

File: docs/generate_docs_for_linux [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+doxygen Doxyfile.linux
```
Comment (breznak) [inline docs/generate_docs_for_linux:2]: ditto, why having two platform-files, when all it does in run a python script (which is available on all platforms). Something like python -m doxygen .... should work here?
@breznak

Comment (johanwiden) [reply docs/generate_docs_for_linux:2]: I have now moved the document generation documentation into docs/README.md
Doxyfile.common: I have specified the INPUT as a set of directories, not a set of files. If one wants to exclude certain files then one will have to specify the files in that directory explicitly.
As regards doxypypy, nothing has happened in https://github.com/Feneric/doxypypy for over a year. Maybe we should look at using something different. But this is at least a start. Without the patched version one can not use Python version > 3.7 neither on Linux nor on Windows.
The FILTER_PATTERNS are not the same for Linux and Windows, and the script files are not identical either. But I will try to merge them.

Comment (johanwiden) [reply docs/generate_docs_for_linux:2]: I fail to see how I can merge the Doxyfiles and filter scripts. The syntax is different in bash versus Windows CMD: Argument specifier is $1 in bash but %1 in CMD.
It is possible to merge the generate_docs_ scripts, but the code becomes contrived:
  https://stackoverflow.com/questions/17510688/single-script-to-run-in-both-windows-batch-and-linux-bash

File: docs/Doxyfile_include_for_windows (hunk 2 above)
Comment (breznak) [reply docs/Doxyfile_include_for_windows:14]: @dkeeney what do you think about this situation at Windows? So far I'm thinking I'd rather use upstream library, and if it does not work for Win..we can generate the docs and publish as part of the CI build anyway?

Comment (johanwiden) [reply docs/Doxyfile_include_for_windows:14]: The upstream library does not work in Linux either. The problem is Python, not Windows. Python evolves, but doxypypy has not kept up with the evolution.

Comment (johanwiden) [reply docs/Doxyfile_include_for_windows:14]: If one specifies that Python 3.7 must be used, and not some later version, then the upstream library can be used.

Comment (dkeeney) [reply docs/Doxyfile_include_for_windows:14]: Lets not lock ourselves into a fixed version of Python.
Is there some other document generator that we could switch to?

Comment (johanwiden) [reply docs/Doxyfile_include_for_windows:14]: With doxygen there is at least one alternative for Python, but it is supposed to be inferior. I will have a look.
There is also Sphinx, which is an alternative to doxygen, and supports both C++ and Python. That one is used in docs/Makefile.
My initial results were better with doxygen, but I will take another look.

Comment (breznak) [reply docs/Doxyfile_include_for_windows:14]: > There is also Sphinx, which is an alternative to doxygen, and supports both C++ and Python 

if it's not too complicated that could be a good choice, given our c++ codebase.

**Comment by johanwiden** [comment]
It turned out that Doxygen now supports Python out of the box.
So I removed most of my changes. It should now be enough just to run doxygen, both on Windows and Linux, and the Doxygen config file is the same for both Windows and LInux.

**Comment by johanwiden** [comment]
GitHub Pages: See
  https://goseeky.wordpress.com/2017/07/22/documentation-101-doxygen-with-github-pages/
step 8. I just submitted addition of new file docs/index.html in this pull request (step 9 in the referenced page).

**Comment by breznak** [comment]
As a followup, it'd be nice to setup our CI builds (see `.github/` ) to to generate the docs and have it accessible online.

**Pull request merged**